### PR TITLE
Minuit2 logging

### PR DESCRIPTION
**Pull request by HDembinski**
This is a draft pull request to document the work in progress and to have a platform for discussion.

The goal of this PR is to add two new verbosity levels for the logging of Minuit2, "Warning" and "Debug", which turn the formerly compile-time options to enable increased logging output into run-time options.

Follows the proposal in #6421

### CODE REVIEW

**Comment by bellenot** [comment]
@phsft-bot build

**Comment by HDembinski** [comment]
@lmoneta The patch is ready for review.

All logging is now done through `MnPrint`, which gained a static method `Log` with a severity level, a prefix for the actual message (usually set to the class name to indicate the origin of the message), and the actual message, which can consist of strings and streamable objects. I added a scope guard to set the prefix. Thanks to RAII, this works also for nested scopes.

Since I was going through all the code anyway, I replaced some c headers with the corresponding cxx headers (e.g. math.h with cmath).

I did not touch Minuit2Minimizer, because it is only used by ROOT and I could not figure out how the logging works there.

**Comment by HDembinski** [comment]
I think the CI bots need to be triggered. The tests succeed when I run locally.

**Comment by bellenot** [comment]
@phsft-bot build

**Comment by Axel-Naumann** [comment]
@phsft-bot build!

**Comment by Axel-Naumann** [comment]
@HDembinski thanks! Could you rewrite this branch so we have meaningful commits to merge? That reduces the chances that one of us hits "Rebase & merge" instead of "Squash & merge" :-)

@lmoneta please have a look?

**Comment by HDembinski** [comment]
Maybe a few more notes to make the review easier, because this is a huge patch.

* The previous code used macros to log things. I found that the macros are not needed, so I replaced all the macros with method calls to MnPrint::Log (or its wrappers MnPrint::Error, MnPrint::Info, ...).
* The old logging made a usual habit of prefix messages with an origin, e.g. when you logged from MnHesse, messages would be prefixed with "MnHesse:". Inside ``MnHesse``, to stay with the example, you might have seen code like this
```c++
MN_INFO_MSG("MnHesse: Some foo happened");
[...]
MN_INFO_MSG("MnHesse: Some bar happened");
[...]
int n = 5;
MN_INFO_MSG("MnHesse: Frobinated", n, "glorps");
```
Setting the prefix by hand was error prone and made the log messages longer and more distracting from the actual code, especially now since all the debug messages should also have this prefix. This was infeasible, so I made a scope guard that sets the prefix once for all logging within the current scope, which considerably simplified the code to
```c++
MnPrintPrefix mnprintprefix("MnHesse"); // appears once at the beginning of a long method
[...]
MnPrint::Info("Some foo happened");
[...]
MnPrint::Info("Some bar happened");
[...]
int n = 5;
MnPrint::Info("Frobinated", n, "glorps");
```
The method ``MnPrint::Info`` automatically uses the prefix set by MnPrintPrefix (which is communicated via a global variable). The scope guard is safe to use. If the prefix is missing, an assert is triggered. Functions called inside the scope all use the same prefix unless they define their own prefix, which then overrides the prefix during the scope of the function call, after which the previous prefix is automatically restored by the scope guard. Each scope guard remembers the previous prefix and restores it in its destructor.
* I generally made sure that the logging messages contain the same information as the previous ones, but I reduced excess whitespace. Some debug messages added several empty lines to the output, which made the logs hard to read.
* In very few isolated cases I removed redundant messages. I did this only when two messages that printed the same information in succession.
* In a few cases, I changed debug output so that variables are printed in one line that were printed in several lines. A log is easier to grep and scroll through when variables are on the same line as the prefix.

**Comment by HDembinski** [comment]
@lmoneta I looked into the log of the failing test, it looks like an unrelated bug in the build. Perhaps rebase to master would fix it?

**Comment by HDembinski** [comment]
I further simplified and improved the design of the new MnPrint in commit https://github.com/root-project/root/pull/6705/commits/c804c2e7f129c8fe7ae5b34a188a45bf6847fc17 and updated the previous note about the design.

**Comment by lmoneta** [comment]
@HDembinski , thank you very much for this nice improvements for the logging in Minuit2!!
I like the way you have implemented the MnPrint class. I think all the needed functionality is there and the PR looks good to me. 
Before merging let me look at it in detail and also try it. After a first look I don't see any particular problems. 
There some errors reported by the Travis CI, but I think those are unrelated to your changes

Cheers

Lorenzo

**Comment by HDembinski** [comment]
Thank you, I am glad. Looking at the output from your side is appreciated. I did checks of the logging output with iminuit, whether it looks good, but I haven't tried this with the TError logging backend.

I am looking forward to use this patch in iminuit. Being able to turn on debugging output at runtime is going to be incredibly useful and it already helped me to understand much better what is called where and when in Minuit. I plan to work on other improvements and this going to be very helpful.

**Comment by HDembinski** [comment]
@lmoneta I looked at the remaining failing tests.

I think I fixed the standalone build and the clang-tidy-modernize issue (CI is still running, we will see).

* clang-tidy-analyzer issues: I don't know how to address these in this PR and would like to propose to postpone fixing these in a separate PR. I looked through the complaints and those seem to be all related to memory management issues in Minuit2 that were there before, but are now uncovered since I touched so many lines of code. I didn't see any complaint related to my actual new logging code. I think a good deal of these issues can be resolved by replacing the MnRefCountedPointer with std::shared_ptr. Writing shared pointers correctly is hard, it is good that we have one in the stdlib now. I would like to do this in a separate PR.

* clang-format issues: I would love to fix these clang-format issues. In fact, I would love to clang-format the whole Minuit2 code right now, but I think this should happen in a separate PR - or perhaps after the patch was fully reviewed. Doing a clang-format on the whole code will probably reveal even more issues for clang-tidy-analyzer...

**Comment by lmoneta** [comment]
Hi Hans,  
 I agree that it is better to fix those issues reported by clang-tidy-analyzer and clang-format in separate PR's.

**Comment by HDembinski** [comment]
@lmoneta Did you have a change to look at the output? Can I help by running some tests in ROOT and posting the output? I have a ROOT build with tests enabled.

**Comment by lmoneta** [comment]
Hi, 
Yes I have tried and noticed some differences in the output produced when fitting in ROOT. Probably the MInimizer interface used in ROOT needs to be adapted for the new loggiing levels. 
 I would prefer doing this and having a fix there ready  before merging this one, so I can commit after merging this PR.

**Comment by HDembinski** [comment]
I see. I changed the formatting in some cases, but I guess you mean that some messages are not shown which should be shown or something is not shown that was shown before? Let me know how I can help.

File: math/minuit2/src/MnSeedGenerator.cxx
Comment (lmoneta) [inline math/minuit2/src/MnSeedGenerator.cxx:75]: Before the function MnPrint::PrintState was printing much less information and all in one line, see:
https://github.com/root-project/root/blob/5134568c209d8288ba38e8d31d45744eb79b0863/math/minuit2/src/MnPrint.cxx#L58

Maybe one could re-use this function. 
Wehn running in debug mode maybe is more interesting having the full state information including parameter values which can be a lot !

File: math/minuit2/src/MnTraceObject.cxx
Comment (lmoneta) [inline math/minuit2/src/MnTraceObject.cxx:23]: Actually this functionality (print in one line) can be used above in MnSeedGenerator.cxx when printing the initial state

File: math/minuit2/src/Minuit2Minimizer.cxx
Comment (lmoneta) [inline math/minuit2/src/Minuit2Minimizer.cxx:429]: Is this line still needed ? Maybe not, since we use later  MnPrint::SetGlobalLevel

File: math/minuit2/src/Minuit2Minimizer.cxx
Comment (lmoneta) [inline math/minuit2/src/Minuit2Minimizer.cxx:433]: Here I think we need to adapt the printLevel of MInimizer to the unse used in MnPrint
Probably we should do something like: 
Minuit2Minimizer::PrintLevel() < 0  --> all message suppressed MnPrint default level = 0
                                                = 0   ->  Only Error message printed  (MnPrint::SetGlobalLevel(1)
                                                = 1   ->  Error + warning message printed: MnPrint::SetGlobalLevel(2)
                                                = 2,3   -> Info message printed : MnPrint::SetGlobalLevel(3)
                                                > 3       -> Debug mode  : MnPrint::SetGlobalLevel(4)

This adaptation is needed to avoid changing output messages drammatically running workflow using MInuit2 with this MinimizerInterface from ROOT (e.g. ROOT fitting, Roofit/RooStats , etc....).
I could eventually apply also after this PR is merged with a following PR

**Review by lmoneta — CHANGES_REQUESTED** [review]

File: math/minuit2/src/Minuit2Minimizer.cxx
Comment (lmoneta) [reply math/minuit2/src/Minuit2Minimizer.cxx:433]: Thank you Hans for this nice improvement in the Minuit2 logging!

I see in Debug mode that a lots of messages are printed. Is it possible to filter the messages on the class that generates them ? That would be useful

Another useful thing would be maybe to have two level of debug verbosity. For example when doing a fit with many parameters I guess you don;t want to print all parameter values and matrix values, because the messages will be unreadable. Maybe in that case only function values could be printed (and all in one line).
Anyway these improvements could also be applied afterwards with an additional PR. 

Cheers

File: math/minuit2/src/MnSeedGenerator.cxx
Comment (HDembinski) [reply math/minuit2/src/MnSeedGenerator.cxx:75]: Ok, I will try to restore this.

File: math/minuit2/src/MnTraceObject.cxx
Comment (HDembinski) [reply math/minuit2/src/MnTraceObject.cxx:23]: Right, I introduced this `Oneline` class to replace `PrintState`, since I want all printing to go through `Info`, `Debug` etc. Ok, I will use that above.

File: math/minuit2/src/Minuit2Minimizer.cxx
Comment (HDembinski) [reply math/minuit2/src/Minuit2Minimizer.cxx:429]: I am not sure, I was a bit confused by the code in Minuit2Minimizer, so I tried to change as little as possible.

A bit off-topic, but do you agree that ultimately our goal is to allow each class to have its own printing level that can be set by users, and to only fall back to a global print level if no custom setting is used?

File: math/minuit2/src/Minuit2Minimizer.cxx
Comment (HDembinski) [reply math/minuit2/src/Minuit2Minimizer.cxx:433]: > I see in Debug mode that a lots of messages are printed. Is it possible to filter the messages on the class that generates them ? That would be useful

I agree that this is useful and on my to-do list for follow-up work. There are several options, please let me know what you think about them:
1) Make all prints one-line, so that one can grep out those of interest, just using the class name.
Pros:
- Simple to implement, simple to understand and simple to use
Cons:
- Lines will become very long in some cases
- Printing vectors and matrices in one line may hurt readability a lot
2) Add a `SetPrintLevel` method to each class, so that the print level can be set individually for each class. If no specific level is chosen, use the global default level.
Pros:
- Very flexible. Can use multi-line printouts.
- Seems to be a standard solution. I have seen such an implementation e.g. in the LHCb software framework.
- Easy to pass down print level recursively, e.g. if MnMigrad is calling MnLineSearch, MnLineSearch can use the print level of MnMigrad.
Cons (sort of):
- Some interface needs to be added in ROOT to allow setting the debug level individually for each class. This needs some thought. In iminuit it is not an issue, because I use the Minuit2 classes directly, but ROOT has this abstract Minimizer interface on top of several minimizers, so it is not clear how to pass the information down to Minuit2. This could perhaps be done with a string based interface: `SetPrintLevel("MnLineSearch", 3);` String-based interfaces are not ideal, but they have the advantage of being perfectly decoupled from the actual implementation below. An alternative is to use an enum instead of a string, but that exposes some details of the Minuit2 class structure.
3) Build a global hash-map into MnPrint which stores individual print levels for each class using a mapping from class name (string) to print level.
Pros:
- A comparably simple extension of the current design.
- All control of print-levels still goes through MnPrint, one does not have to set print-levels individually on each class with a method on that class.
Cons:
- String-based interfaces are brittle, typos can only be detected at run-time.
- Like in 2) some interface in ROOT is needed to allow users to set the print level.
- Passing-down print levels is not so straight forward in this design.

I suppose the best short-term solution is 3). We should decide whether we generally want classes that are internally called by other classes to inherit the print level of the caller. If that's desired, then solution 2) is cleaner. If not, then I would go for 3). I cannot predict at this point which design will be better in practice.

If we decide to go for 3), should I make the changes in this PR or implement that in a follow up PR?

In general, I am also very interested in more fine-grained control on the printing also for iminuit, although I would also like to keep an option to "show me everything", because it really helped me a lot to understand what Migrad actually does and in what order.

Comment (lmoneta) [reply math/minuit2/src/Minuit2Minimizer.cxx:433]: Hi Hans, 
Thank you very much for your detailed analysis on how to implement a fine grained printing in Minuit2. 
My preference would be probably to go for option (3) that is maybe less clean, but also more practical to implement. 
However I see that option (2) has the advantage to have a recursivw print level. 

Being practical,  we would not need a detailed level for every single class, but for maybe some group of class implementing a specific algorithm. For example you would like to debug the iterations in the VariableMetric algorithm (MnMigrad) or the computation of numerical derivatives or the line search algorithm. 
This could be controlled maybe easier with a global hash map as you suggest.
There is often the use case you want to debug a lower level class (e.g. MnLineSearch) but not debugging an higher level (MnMigrad) or the opposite (e.g. debugging MnMinos but not MnMigrad). 
Of course there is also the case you want to see everything. 
A simpler solution is also having multiple debug levels which apply to all Minuit. 

Anyway, I agree it is better to merge first this PR and do improvements eventually in a following one.  

For this PR, I would add just the fix for the Info message in MnSeedGenerator. I can then commit later a subsequent PR to adapt the Minuit2Minimizer setting of the Print Level. 

Thank you !

Comment (HDembinski) [reply math/minuit2/src/Minuit2Minimizer.cxx:433]: Ok, will do! Thank you for the discussion! What you said gave me some food for thought.

I think we both prefer option 3) then, but this approach should still make it easy to distinguish, staying with your example, MnLineSearch called by MnMigrad from MnLineSearch called by MnMinos. This means that the MnPrint instances need to know when they are nested.  I think this can be achieved with a minor modification of the current design at the cost of again manipulating global state.

Since MnPrint is already a scope guard, it could modify a global prefix. Pseudo-code:
```c++
{
  // in MnMigrad scope
  MnPrint print("MnMigrad"); // adds MnMigrad to global prefix
 
  print.Info("foo"); // prints MnPrint: foo
  // ... at some point an instance of MnLineSearch is created
  {
     // in MnLineSearch scope
     MnPrint print("MnLineSearch"); // adds MnLineSearch to global prefix
     // ...
     print.Info("bar"); // print MnPrint::MnLineSearch: bar
     // end of scope, remove MnLineSearch from global prefix
  }
  // end of scope, remove MnMigrad from global prefix
}
```
The only issue is to make this thread-safe. The global prefix must be a thread-local variable.

File: math/minuit2/src/MnSeedGenerator.cxx
Comment (HDembinski) [reply math/minuit2/src/MnSeedGenerator.cxx]: ok, now implemented

**Comment by HDembinski** [comment]
I addition to the fix for the one-line printing, I would like to get one last commit in, https://github.com/root-project/root/pull/6705/commits/bcd7adeddf2997feb588ec1c61a67e0247cdb92f. I think the warning levels need to be tweaked for some messages, I did this here for VariableMetricBuilder. I changed two warnings into info messages, because the reported issue is not so severe that it warrants a warning, and it is something that happens fairly often in my experience. On the other hand, I promoted a warning to an error message. If VariableMetricBuilder cannot run at all because the initial EDM is negative, I would consider this an error.

File: math/minuit2/src/Minuit2Minimizer.cxx
Comment (HDembinski) [reply math/minuit2/src/Minuit2Minimizer.cxx:433]: In https://github.com/root-project/root/pull/6705/commits/a6968f4a79a3728385d1461400920e03d829e569, I implemented this idea and I made the global variables related to MnPrint thread_local (which I should have done anyway).

I like these stacking prefixes, because they show how algorithm calls are nested. The downside is that the prefix gets rather long (which I find acceptable). An example from iminuit:
```
I MnApplication:MnSeedGenerator: Initial state: FCN =                 0 Edm =                 0 NCalls =      9
I MnApplication:ModularFunctionMinimizer:VariableMetricBuilder: Start iterating until Edm is < 0.0002 with call limit = 420
I MnApplication:ModularFunctionMinimizer:VariableMetricBuilder:    0 - FCN =                 0 Edm =                 0 NCalls =      9
W MnApplication:ModularFunctionMinimizer:VariableMetricBuilder:MnHesse:MnPosDef: Matrix forced pos-def by adding to diagonal 0.002
I MnApplication:ModularFunctionMinimizer:VariableMetricBuilder: After Hessian
I MnApplication:ModularFunctionMinimizer:VariableMetricBuilder:    1 - FCN =                 0 Edm =                 0 NCalls =     19
```

**Comment by lmoneta** [comment]
Hi Hans, 
Thank you for the update. 
Actually having the info message with all the nested calls is maybe making the line too verbose, especially when calling from ROOT. I am not sure everybody will like this. Here is for example what I am getting:
```
Minuit2Minimizer: Minimize with max-calls 1000000 convergence for edm < 0.001 strategy 1
Info in <Minuit2>: Minuit2Minimizer::Minimize:MnSeedGenerator: Initial state: FCN =              3616 Edm =       6305.439852 NCalls =      9
Info in <Minuit2>: Minuit2Minimizer::Minimize:ModularFunctionMinimizer:VariableMetricBuilder: Start iterating until Edm is < 2e-06 with call limit = 1000000
Info in <Minuit2>: Minuit2Minimizer::Minimize:ModularFunctionMinimizer:VariableMetricBuilder:    0 - FCN =              3616 Edm =       6305.439852 NCalls =      9
Info in <Minuit2>: Minuit2Minimizer::Minimize:ModularFunctionMinimizer:VariableMetricBuilder:    1 - FCN =       13.13450166 Edm =      0.7502570511 NCalls =     16
Info in <Minuit2>: Minuit2Minimizer::Minimize:ModularFunctionMinimizer:VariableMetricBuilder:    2 - FCN =       12.64308061 Edm =    0.001706300191 NCalls =     22
Info in <Minuit2>: Minuit2Minimizer::Minimize:ModularFunctionMinimizer:VariableMetricBuilder:    3 - FCN =       11.13331108 Edm =      0.4006477323 NCalls =     34
Info in <Minuit2>: Minuit2Minimizer::Minimize:ModularFunctionMinimizer:VariableMetricBuilder:    4 - FCN =       10.43729238 Edm =       1.767931142 NCalls =     41
Info in <Minuit2>: Minuit2Minimizer::Minimize:ModularFunctionMinimizer:VariableMetricBuilder:    5 - FCN =       6.552494076 Edm =     0.03027421909 NCalls =     51

```

Maybe we should add a compiler option for this, whichcould be enabled for Iminuit ?

**Comment by lmoneta** [comment]
And I agree with your changes from Warnings to Error and Warnings to Info in bcd7ade

**Comment by HDembinski** [comment]
I would again prefer to have a run-time option and I think I have a solution that satisfies everyone. This is implemented in my latest patch https://github.com/root-project/root/pull/6705/commits/365bd46563c025a5341b48e6d199b2aa4be76c4e.

I think we need to collect the prefix stack if we want to be able to filter out, for example, messages from MnLineSearch that happen inside MnMigrad, but not those that happen inside MnHesse. That's why I think the prefix stack should stay. It is also fast, since only a couple of pointers are copied around. However, there is no need to show the full stack of prefixes. By default, MnPrint could just show the prefix from the innermost scope. I added the static method `MnPrint::ShowPrefixStack(bool yes)` to toggle showing the full prefix or not. The default is off so that the messages are not too long.

I also now implemented filtering of messages based on the value of the prefix stack. This is done with strings, using `MnPrint::AddFilter` and `MnPrint::ClearFilter` to reset the filters. Filtering works like `grep` on the full prefix, the message is shown if any of the filter strings match. To only see messages from `MnHesse`, one can do:
```c++
MnPrint::AddFilter("MnHesse");
```
To only see messages from `MnHesse` called from VariableMetrixBuilder, one can do:
```c++
MnPrint::AddFilter("VariableMetricBuilder:MnHesse");
```
To only match the beginning of the prefix, use "^". For example,
```c++
MnPrint::AddFilter("^MnHesse");
```
only matches direct execution of MnHesse, but not MnHesse called by MnMigrad.

Filtering is not implemented in a very efficient way to keep the code simple, but the performance drop is opt-in. If no filters are used, then there is no performance loss.

**Comment by HDembinski** [comment]
@lmoneta Feel free to ask me to revert this if you don't like the filtering. I just threw this in because we were discussing follow-up work and I wanted to show how the prefix stack can be used for filtering, even if it is not displayed.

**Comment by lmoneta** [comment]
Hi Hans, 
 I think your idea of filtering and controlling the prefix stack at run time is great!
I like it.
Let me try and run a CI on the latest commit and then I think we can finally approve this PR
Thank you for this nice improvement!

**Comment by lmoneta** [comment]
@phsft-bot build !

**Comment by lmoneta** [comment]
@HDembinski Hi , 
there are some compiler errors in MnPrint and a warning in FumiliGradientCalculator.cxx

**Comment by lmoneta** [comment]
When running from ROOT the tutorial `tutorials/fit/NumericalMinimization.C`
I get this crash when using a debug print level (>3)
```
Minuit2Minimizer: Minimize with max-calls 1000000 convergence for edm < 0.001 strategy 1
Info in <Minuit2>: MnSeedGenerator 2 free parameters, FCN pointer 0x7ffd440f3d30
Info in <Minuit2>: InitialGradientCalculator Calculating initital gradient at point 
     -19.29441224
      18.75662556
Info in <Minuit2>: InitialGradientCalculator Computed initial gradient for parameter x value -19.2944 [ -0.01 , 0.01 ] dirin 0.01 grd 200 g2 20000
Info in <Minuit2>: InitialGradientCalculator Computed initial gradient for parameter y value 18.7566 [ -0.01 , 0.01 ] dirin 0.01 grd 200 g2 20000
Info in <Minuit2>: Numerical2PGradientCalculator Assumed precision eps 8.88178e-16 eps2 5.96046e-08
Info in <Minuit2>: Numerical2PGradientCalculator Calculating gradient around value 1.24979e+07 at point 
     -19.29441224
      18.75662556
Info in <Minuit2>: Numerical2PGradientCalculator cycle 0 x -19.29441223852 step 0.01 f1 12470627.44365 f2 12525195.60355 grd -2728407.995034 g2 439228.5821213
Info in <Minuit2>: Numerical2PGradientCalculator cycle 1 x -19.29441223852 step 0.003683478230861 f1 12487842.51291 f2 12507942.5709 grd -2728407.327973 g2 439228.5648472
Info in <Minuit2>: Numerical2PGradientCalculator Parameter x Gradient -2728407.327973 g2 439228.5648472 step 0.003683478230861

 *** Break *** segmentation violation



===========================================================
There was a crash.
This is the entire stack trace of all threads:
===========================================================

Thread 16 (Thread 0x7fd8cb1bc700 (LWP 14166)):
#0  0x00007fd8d2568113 in ?? () from /usr/lib/x86_64-linux-gnu/libgomp.so.1
#1  0x00007fd8d2e38c5b in ROOT::Minuit2::Numerical2PGradientCalculator::operator()(ROOT::Minuit2::MinimumParameters const&, ROOT::Minuit2::FunctionGradient const&) const [clone ._omp_fn.0] () from /home/moneta/rootgit/root_dev3-build/lib/libMinuit2.so
#2  0x00007fd8d256596e in ?? () from /usr/lib/x86_64-linux-gnu/libgomp.so.1
#3  0x00007fd8e630a6db in start_thread (arg=0x7fd8cb1bc700) at pthread_create.c:463
#4  0x00007fd8e6efca3f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

```

**Comment by HDembinski** [comment]
Ok, I am currently looking into those.

**Comment by HDembinski** [comment]
@lmoneta I fixed the error and the warning. I tried `tutorials/fit/NumericalMinimization.C` with the fixed code on my computer and its works. I changed the line 55 to increase the print level to 4.

**Comment by lmoneta** [comment]
@phsft-bot build!

**Comment by lmoneta** [comment]
Thank you for the fixes !
I am still having the crash on my machine with debug print level. Maybe is due to some bad configuration on my side. Let me check this

**Comment by lmoneta** [comment]
Hi 
The crash I observed seems to come when Minuit2 is compiled with omp support (i.e. when _OPENMP is defined in the 
Numerical2PGradientCalculator. 
Maybe we should add the debug printing in the for -loop only when _OPENMP is not defined

**Comment by lmoneta** [comment]
I have tried adding the pragma omp critical (see below)  but still I am having the crash. 
```
#ifdef _OPENMP
#pragma omp critical
#endif
     {
      print.Debug([&](std::ostream &os) {
         const int pr = os.precision(13);
         const int iext = Trafo().ExtOfInt(i);
         os << "Parameter " << Trafo().Name(iext) << " Gradient " << grd(i) << " g2 " << g2(i) << " step " << gstep(i);
         os.precision(pr);
      });
     }
```

so for the time being I would not use print.Debug when _OPENMP is defined

**Comment by lmoneta** [comment]
Investigating more the crash is happening at  MnPrint.cxx:150, from this line: 
```
os << gPrefixStack.back();
```

If I call `MnPrint::ShowPrefixStack(true)` then is ok.

**Comment by HDembinski** [comment]
@lmoneta Thanks for investigating. I am a bit confused now. If this crash only happens when OMP is enabled, then it is probably related to my use of thread_local variables. I don't have much experience with what can go wrong there. Is there an easy way for me to reproduce this on my computer?

**Comment by HDembinski** [comment]
If calling gPrefixStack.back() segfaults, then the gPrefixStack is empty when it is called. The code assumes that gPrefixStack is never empty when StreamPrefix is called, which under normal circumstances is indeed impossible, one cannot call logging without creating a MnPrint object and that action adds the first prefix to the gPrefixStack. All that assumes single-thread execution, of course. I seems that in the multi-threaded case there are either race conditions despite the use of thread-local variables, or thread-local variables work differently from what I assumed. I assumed that when a new thread is started from the main thread, the current value from the thread-local variable in the main thread is copied to the new thread. If the thread-local variable in the new thread is default initialized instead, then that would explain the crash.

**Comment by HDembinski** [comment]
I read up on this and it is indeed so that the thread-local variable is default initialized in each thread, therefore the gPrefixStack is empty. I moved the debug print outside of the threads, it now happens after the threads are joined. I hope this fixes the issue.

I now compiled ROOT with minuit2_omp=ON and I don't get crashes, but some tests fail.
```
	164 - minuit2_testMinimizer (Failed)
	227 - test-stressroostats-minuit2 (Failed)
	414 - tutorial-fit-minuit2FitBench (Failed)
	416 - tutorial-fit-minuit2GausFit (Failed)
```
Is this normal?

Some notes on the other two patches:
* https://github.com/root-project/root/pull/6705/commits/83af202e0940423c2ba0fcdb0e091d861d5f517f just fixes a typo that I noticed.
* https://github.com/root-project/root/pull/6705/commits/de1d979c1c9d194852237466dbf2ba04322f15df I noticed that MnApplication only prints the debug output when the user does not pass an analytical gradient, so I fixed that, too,

**Comment by lmoneta** [comment]
Hi, 
Thank you for the fix. I do not have anymore that crash in gPrefixStack. 
Yes, I can reproduce those failures of the other tests running with multi-thread and openMP. This is most likely caused by non-thread safe implementations of the likelihood functions used for the fitting. It is something needing a fix, but it is outside of the scope of this PR. 

I think we are ready to merge this one. Let me run the CI first.

**Comment by lmoneta** [comment]
@phsft-bot build

**Comment by lmoneta** [comment]
@phsft-bot build with flags -Dminuit2=On -Dminuit2_omp=On

File: math/minuit2/src/Numerical2PGradientCalculator.cxx
Comment (lmoneta) [inline math/minuit2/src/Numerical2PGradientCalculator.cxx:200]: This prints the message in a single line in a no readable message. 
What about changing to something like this that is more readable: 
```
  print.Debug([&](std::ostream& os) {
      const int pr = os.precision(13);
      os << std::endl;
      os << std::setw(14) << "Parameter" << std::setw(14) << "Gradient" << std::setw(14) << "g2 "
         << std::setw(14) << "step" << std::endl;
      for (int i = 0; i < int(n); i++) {
         const int iext = Trafo().ExtOfInt(i);
         os << std::setw(14) << Trafo().Name(iext) << " " << grd(i) << " " << g2(i) << " " << gstep(i) << std::endl;
      }
      os.precision(pr);
   });
```

similar could be done for other messages (see for example print.Debug above)

File: math/minuit2/src/Numerical2PGradientCalculator.cxx
Comment (lmoneta) [inline math/minuit2/src/Numerical2PGradientCalculator.cxx:170]: It would be nice to have these lines also availablewhen _OPENMP is enabled.
I was trying to debug the failing tests seen in OPENMP
Unfortunately if enabled, it is still crashing in MnPrint.Debug as before.
Suggested changes to reproduce it and improve printout quality
```
#ifdef _OPENMP
#pragma omp critical
#endif
         {
            if (i == 0 && j == 0) {
               print.Debug([&](std::ostream &os) {
                  os << std::setw(10) << "parameter" << std::setw(6) << "cycle" << std::setw(15) << "x" << std::setw(15)
                     << "step" << std::setw(15) << "f1" << std::setw(15) << "f2" << std::setw(15) << "grd"
                     << std::setw(15) << "g2" << std::endl;
               });
            }
            print.Debug([&](std::ostream &os) {
               const int pr = os.precision(13);
               const int iext = Trafo().ExtOfInt(i);
               os << std::setw(10) << Trafo().Name(iext) << std::setw(5) << j << "  " << x(i) << " " << step << " "
                  << fs1 << " " << fs2 << " " << grd(i) << " " << g2(i) << std::endl;
               os.precision(pr);
            });
         }
```

File: math/minuit2/src/Numerical2PGradientCalculator.cxx
Comment (HDembinski) [reply math/minuit2/src/Numerical2PGradientCalculator.cxx:200]: Ok. I have no strong opinion on how messages should be formatted, I always assumed that the formatting will be changed in follow-up PRs. Because we have the filters now, there is no reason left to make messages one-liners.

Comment (lmoneta) [reply math/minuit2/src/Numerical2PGradientCalculator.cxx:200]: OK, I agree. Let's do the formatting in following PR's. But anyway a `std::endl` was missing

File: math/minuit2/src/Numerical2PGradientCalculator.cxx
Comment (HDembinski) [reply math/minuit2/src/Numerical2PGradientCalculator.cxx:170]: If you want to print inside the thread, you need to create a new MnPrint instance inside the thread. Then there should be no crash. You can create one inside the braced scope { } here.

File: math/minuit2/src/Numerical2PGradientCalculator.cxx
Comment (HDembinski) [reply math/minuit2/src/Numerical2PGradientCalculator.cxx:200]: I implemented the change as you suggested.

File: math/minuit2/src/Numerical2PGradientCalculator.cxx
Comment (HDembinski) [reply math/minuit2/src/Numerical2PGradientCalculator.cxx]: I added the code as described.

**Comment by lmoneta** [comment]
@phsft-bot build with flags -Dminuit2=On

File: math/minuit2/src/Numerical2PGradientCalculator.cxx
Comment (HDembinski) [reply math/minuit2/src/Numerical2PGradientCalculator.cxx]: The core problem is really the use of global variables with threads. The is no golden way and no simple solution, we don't know where the thread is formed. MnPrint should work if a whole MnApplication is run in parallel and it should also work when a thread is started locally in Numerical2PCalculator. It is difficult to design something that works in both cases.

**Review by lmoneta — APPROVED** [review]
Thank you for the latest update and the fix for MT. 
It looks good to me know!
I will merge it.
Thank you for the contribution!

**Comment by HDembinski** [comment]
@lmoneta Big thanks! Sorry for the huge patch, reviewing this was surely not easy.

**Comment by lmoneta** [comment]
Thanks to you for this useful addition to Minuit2. 
It is now merged in the master!
Cheers

**Pull request merged**
